modules/loader/source/relocations.cpp

```diff
@@ -832,6 +839,51 @@ bool resolveRISCV(const loader::Relocation &r, loader::ElfMap &map,
       writeITypeImm(getLo12(relative_value), relocation_address + 4);
       break;
     }
+    case R_RISCV_BRANCH: {
+      // R_RISCV_BRANCH immediate is for B-Type instructions and is split
+      // across 12 bits spread across the instruction and skips the bottom bit
+      // see conditional branches in the isa spec
+      int64_t signed_relative_value = static_cast<int64_t>(relative_value);
+      uint64_t abs_val =
+          signed_relative_value < 0 ? -signed_relative_value : relative_value;
+      if (abs_val & (~0x8FFE)) {
```

Comment:
This check does not look right. The constant used here doesn't have consecutive bits set, and by checking absolute value, the largest valid negative value is rejected.

We might be able to fix this very simply:  in `getHi20`, we do
```
  // Sign extends a number in the bottom 'b' bits of a 64-bit number back up to
  // a 64-bit number.
  auto signExtendN = [](uint64_t val, unsigned b) {
    CARGO_ASSERT(b > 0, "bit-width cannot be zero");
    CARGO_ASSERT(b <= 64, "bit-width out of range");
    return static_cast<int64_t>(val << (64 - b)) >> (64 - b);
  };
  ...
  // Check that the value, when sign-extended back up to a 64-bit number (e.g.,
  // through LUI) will not lose bits.
  if ((addr >> 12) != signExtendN(addr >> 12, 20)) {
```
If we make `signExtendN` a function rather than just a helper lambda, we could do the check that way here too.

Reply:
This looks good, thanks. There's still the previous version of the code commented out which I assume was an accident.

To make sure it doesn't get missed, commenting here as well: could you add a check that the instruction after an `ALIGN` relocation is suitably aligned?

Reply:
Sure yes. Will tidy up before committing